src/exercises/day-3/safe-ffi-wrapper.rs

```diff
@@ -25,43 +25,46 @@ mod ffi {
         _marker: core::marker::PhantomData<(*mut u8, core::marker::PhantomPinned)>,
     }
 
-    // Layout as per readdir(3) and definitions in /usr/include/x86_64-linux-gnu.
+    // Layout according to the Linux man page for readdir(3), where ino_t and
+    // off_t are resolved according to the definitions in
+    // /usr/include/x86_64-linux-gnu/{sys/types.h, bits/typesizes.h}.
     #[cfg(not(target_os = "macos"))]
     #[repr(C)]
     pub struct dirent {
-        pub d_ino: c_long,
-        pub d_off: c_ulong,
+        pub d_ino: c_ulong,
+        pub d_off: c_long,
```

Comment:
Just out of curiosity, how did you find this? I traced this way back and it's super likely that I misread something when pretending to a be a C preprocessor in my head :smile:

Reply:
https://android.googlesource.com/platform/prebuilts/gcc/linux-x86/host/x86_64-linux-glibc2.15-4.8/+/a5ef72f/sysroot/usr/include/x86_64-linux-gnu/sys/types.h has off_t --> __off_t / __off64_t and something similar for ino_t.

I looked at the includes, and had a lucky guess that bits/types.h would be useful.

https://android.googlesource.com/platform/prebuilts/gcc/linux-x86/host/x86_64-linux-glibc2.15-4.8/+/a5ef72f/sysroot/usr/include/x86_64-linux-gnu/bits/types.h has the decoder ring for __ULONGWORD_TYPE / __SLONGWORD_TYPE / __SLONGWORD_TYPE / etc., which show up below, and suggests that __off_t / __off64_t etc. match __OFF_T_TYPE  / __OFF64_T_TYPE -- I assumed that __STD_TYPE is some form of typedef.

https://android.googlesource.com/platform/prebuilts/gcc/linux-x86/host/x86_64-linux-glibc2.15-4.8/+/a5ef72f/sysroot/usr/include/x86_64-linux-gnu/bits/typesizes.h defines __OFF_T_TYPE, __OFF64_T_TYPE, etc. in terms of __ULONGWORD_TYPE etc. introduced  in types.h.

At a higher level, while looking at the macOS crash on CI, I stumbled upon something like https://stackoverflow.com/a/48169793/537046 which told me POSIX mandates that `ino_t` is unsigned. This was my hint that I should investigate :)

Reply:
Also, those links are what Google search gave me when I asked for ino_t and off_t on Linux.

Reply:
Awesome, thanks!